### PR TITLE
Register restart after system reboot, #2154

### PR DESCRIPTION
**Pull request by studentmain**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

- [x] [Searched](https://github.com/shadowsocks/shadowsocks-windows/search?q=is%3Apr&type=Issues) for similar pull requests
- [x] Compiled the code with Visual Studio

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New feature

---

### Description of your *pull request* and other information

在系统重启或更新后自启。在我的机器上可用，可能需要更多测试。
Restart program after system reboot/update. Works on my machine, may need more test.

### CODE REVIEW

File: shadowsocks-csharp/Controller/System/AutoStartup.cs [hunk 1]
```diff
@@ -89,13 +96,47 @@ public static bool Check()
...
+
+        enum ApplicationRestartFlags
+        {
+            RESTART_NO_CRASH = 1,
+            RESTART_NO_HANG = 2,
+            RESTART_NO_PATCH = 4,
+            RESTART_NO_REBOOT = 8,
+        }
+
+        // regist restart after system reboot/update
```
Comment (celeron533) [inline shadowsocks-csharp/Controller/System/AutoStartup.cs]: regist --> register

File: shadowsocks-csharp/Controller/System/AutoStartup.cs [hunk 2]
```diff
@@ -89,13 +96,47 @@ public static bool Check()
...
+        public static void RegisterForRestart(bool regist)
+        {
+            // requested regist and not autostart
+            if (regist && !Check())
+            {
+                // param 1 is process command param
+                RegisterApplicationRestart(null, (int)ApplicationRestartFlags.RESTART_NO_CRASH | (int)ApplicationRestartFlags.RESTART_NO_HANG);
```
Comment (celeron533) [inline shadowsocks-csharp/Controller/System/AutoStartup.cs]: Better to set current process's parameter (although it is not implemented right now) as param 1

Comment (studentmain) [reply shadowsocks-csharp/Controller/System/AutoStartup.cs]: doing

Comment (studentmain) [reply shadowsocks-csharp/Controller/System/AutoStartup.cs]: done

Comment (studentmain) [reply shadowsocks-csharp/Controller/System/AutoStartup.cs]: Should I use [this](https://docs.microsoft.com/en-us/dotnet/api/system.environment.getcommandlineargs) to get argv?

**Comment by sxyandapp** [comment]
看似人性化，实则很流氓的功能

**Pull request merged**